File: Source/OldCode/Archive/2005 TVA Code Library/Version/2.0/Local/Source/Tva.Phasors/FrequencyDefinitionBase.vb
'*******************************************************************************************************
'  FrequencyDefinitionBase.vb - Frequency and df/dt value definition base class
'  Copyright © 2005 - TVA, all rights reserved - Gbtc
'
'  Build Environment: VB.NET, Visual Studio 2005
'  Primary Developer: J. Ritchie Carroll, Operations Data Architecture [TVA]
'      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
'       Phone: 423/751-2827
'       Email: jrcarrol@tva.gov
'
'  Code Modification History:
'  -----------------------------------------------------------------------------------------------------
'  02/18/2005 - J. Ritchie Carroll
'       Initial version of source generated
'
'*******************************************************************************************************

Imports System.Runtime.Serialization
Imports System.ComponentModel
Imports Tva.Phasors.Common

''' <summary>This class represents the common implementation of the protocol independent definition of a frequency and df/dt value.</summary>
<CLSCompliant(False), Serializable()> _
Public MustInherit Class FrequencyDefinitionBase

    Inherits ChannelDefinitionBase
    Implements IFrequencyDefinition

    Private m_dfdtScale As Int32
    Private m_dfdtOffset As Single

    Protected Sub New()
    End Sub

    Protected Sub New(ByVal info As SerializationInfo, ByVal context As StreamingContext)

        MyBase.New(info, context)

        ' Deserialize frequency definition
        m_dfdtScale = info.GetInt32("dfdtScale")
        m_dfdtOffset = info.GetSingle("dfdtOffset")

    End Sub

    Protected Sub New(ByVal parent As IConfigurationCell)

        MyBase.New(parent)

    End Sub

    Protected Sub New(ByVal parent As IConfigurationCell, ByVal label As String, ByVal scale As Int32, ByVal offset As Single, ByVal dfdtScale As Int32, ByVal dfdtOffset As Single)

        MyBase.New(parent, 0, label, scale, offset)

        m_dfdtScale = dfdtScale
        m_dfdtOffset = dfdtOffset

    End Sub

    Protected Sub New(ByVal parent As IConfigurationCell, ByVal binaryImage As Byte(), ByVal startIndex As Int32)

        MyBase.New(parent, binaryImage, startIndex)

    End Sub

    ' Derived classes are expected to expose a Public Sub New(ByVal frequencyDefinition As IFrequencyDefinition)
    Protected Sub New(ByVal frequencyDefinition As IFrequencyDefinition)

        MyClass.New(frequencyDefinition.Parent, frequencyDefinition.Label, frequencyDefinition.ScalingFactor, _
            frequencyDefinition.Offset, frequencyDefinition.DfDtScalingFactor, frequencyDefinition.DfDtOffset)

    End Sub

    Public Overrides ReadOnly Property DataFormat() As DataFormat
        Get
            Return Parent.FrequencyDataFormat
        End Get
    End Property

    Public Overridable ReadOnly Property NominalFrequency() As LineFrequency Implements IFrequencyDefinition.NominalFrequency
        Get
            Return Parent.NominalFrequency
        End Get
    End Property

    <EditorBrowsable(EditorBrowsableState.Never)> _
    Public Overrides Property Index() As Integer
        Get
            Return MyBase.Index
        End Get
        Set(ByVal value As Integer)
            MyBase.Index = value
        End Set
    End Property

    Public Overrides Property Offset() As Single
        Get
            Return Convert.ToSingle(Parent.NominalFrequency)
        End Get
        Set(ByVal value As Single)
            Throw New NotSupportedException("Frequency offset is read-only - it is determined by nominal frequency specified in containing condiguration cell")
        End Set
    End Property

    Public Overridable Property DfDtOffset() As Single Implements IFrequencyDefinition.DfDtOffset
        Get
            Return m_dfdtOffset
        End Get
        Set(ByVal value As Single)
            m_dfdtOffset = value
        End Set
    End Property

    Public Overridable Property DfDtScalingFactor() As Int32 Implements IFrequencyDefinition.DfDtScalingFactor
        Get
            Return m_dfdtScale
        End Get
        Set(ByVal value As Int32)
            m_dfdtScale = value
        End Set
    End Property

    Public Overrides Sub GetObjectData(ByVal info As System.Runtime.Serialization.SerializationInfo, ByVal context As System.Runtime.Serialization.StreamingContext)

        MyBase.GetObjectData(info, context)

        ' Serialize frequency definition
        info.AddValue("dfdtScale", m_dfdtScale)
        info.AddValue("dfdtOffset", m_dfdtOffset)

    End Sub

    Public Overrides ReadOnly Property Attributes() As Dictionary(Of String, String)
        Get
            Dim baseAttributes As Dictionary(Of String, String) = MyBase.Attributes

            With baseAttributes
                .Add("df/dt Offset", DfDtOffset)
                .Add("df/dt Scaling Factor", DfDtScalingFactor)
            End With

            Return baseAttributes
        End Get
    End Property

End Class

File: Source/OldCode/Archive/2007 Phasor Protocols/PhasorProtocols/FrequencyDefinitionBase.cs
//*******************************************************************************************************
//  FrequencyDefinitionBase.cs
//  Copyright © 2009 - TVA, all rights reserved - Gbtc
//
//  Build Environment: C#, Visual Studio 2008
//  Primary Developer: James R Carroll
//      Office: PSO TRAN & REL, CHATTANOOGA - MR BK-C
//       Phone: 423/751-4165
//       Email: jrcarrol@tva.gov
//
//  Code Modification History:
//  -----------------------------------------------------------------------------------------------------
//  02/18/2005 - James R Carroll
//       Generated original version of source code.
//
//*******************************************************************************************************

using System;
using System.ComponentModel;
using System.Collections.Generic;
using System.Runtime.Serialization;

namespace PCS.PhasorProtocols
{
    /// <summary>
    /// Represents the common implementation of the protocol independent definition of a <see cref="IFrequencyValue"/>.
    /// </summary>
    [Serializable()]
    public abstract class FrequencyDefinitionBase : ChannelDefinitionBase, IFrequencyDefinition
    {
        #region [ Members ]

        // Fields
        private uint m_dfdtScale;
        private double m_dfdtOffset;

        #endregion

        #region [ Constructors ]

        /// <summary>
        /// Creates a new <see cref="FrequencyDefinitionBase"/>.
        /// </summary>
        /// <param name="parent">The <see cref="IConfigurationCell"/> parent of this <see cref="FrequencyDefinitionBase"/>.</param>
        protected FrequencyDefinitionBase(IConfigurationCell parent)
            : base(parent)
        {
        }

        /// <summary>
        /// Creates a new <see cref="FrequencyDefinitionBase"/> from specified parameters.
        /// </summary>
        /// <param name="parent">The <see cref="IConfigurationCell"/> parent of this <see cref="FrequencyDefinitionBase"/>.</param>
        /// <param name="label">The label of this <see cref="FrequencyDefinitionBase"/>.</param>
        /// <param name="scale">The integer scaling value of this <see cref="FrequencyDefinitionBase"/>.</param>
        /// <param name="dfdtScale">The df/dt scaling value of this <see cref="FrequencyDefinitionBase"/>.</param>
        /// <param name="dfdtOffset">The df/dt offset of this <see cref="FrequencyDefinitionBase"/>.</param>
        protected FrequencyDefinitionBase(IConfigurationCell parent, string label, uint scale, uint dfdtScale, double dfdtOffset)
            : base(parent, label, scale, 0.0D)
        {
            m_dfdtScale = dfdtScale;
            m_dfdtOffset = dfdtOffset;
        }

        /// <summary>
        /// Creates a new <see cref="FrequencyDefinitionBase"/> from serialization parameters.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> with populated with data.</param>
        /// <param name="context">The source <see cref="StreamingContext"/> for this deserialization.</param>
        protected FrequencyDefinitionBase(SerializationInfo info, StreamingContext context)
            : base(info, context)
        {
            // Deserialize frequency definition
            m_dfdtScale = info.GetUInt32("dfdtScale");
            m_dfdtOffset = info.GetDouble("dfdtOffset");
        }

        #endregion

        #region [ Properties ]

        /// <summary>
        /// Gets the <see cref="PhasorProtocols.DataFormat"/> of this <see cref="FrequencyDefinitionBase"/>.
        /// </summary>
        public override DataFormat DataFormat
        {
            get
            {
                return Parent.FrequencyDataFormat;
            }
        }

        /// <summary>
        /// Gets the nominal <see cref="LineFrequency"/> of this <see cref="FrequencyDefinitionBase"/>.
        /// </summary>
        /// <remarks>
        /// Value returned is the <see cref="IConfigurationCell.NominalFrequency"/> and is exposed here just for convenience.
        /// </remarks>
        public virtual LineFrequency NominalFrequency
        {
            get
            {
                return Parent.NominalFrequency;
            }
        }

        /// <summary>
        /// Gets or sets the index of this <see cref="FrequencyDefinitionBase"/>.
        /// </summary>
        /// <remarks>
        /// Phasor protocols only define one frequency measurement per device.
        /// </remarks>
        [EditorBrowsable(EditorBrowsableState.Never)]
        public override int Index
        {
            get
            {
                return base.Index;
            }
            set
            {
                base.Index = value;
            }
        }

        /// <summary>
        /// Gets or sets the offset of this <see cref="FrequencyDefinitionBase"/>.
        /// </summary>
        /// <remarks>
        /// Offset for frequency values will always be the nominal frequency as defined in parent configuration cell; assigning a value is not supported.
        /// </remarks>
        /// <exception cref="NotSupportedException">Frequency offset is read-only; value is determined by nominal frequency specified in containing configuration cell.</exception>
        public override double Offset
        {
            get
            {
                return (double)Parent.NominalFrequency;
            }
            set
            {
                throw new NotSupportedException("Frequency offset is read-only; value is determined by nominal frequency specified in containing configuration cell");
            }
        }

        /// <summary>
        /// Gets or sets the df/dt offset of this <see cref="FrequencyDefinitionBase"/>.
        /// </summary>
        public virtual double DfDtOffset
        {
            get
            {
                return m_dfdtOffset;
            }
            set
            {
                m_dfdtOffset = value;
            }
        }

        /// <summary>
        /// Gets or sets the df/dt scaling value of this <see cref="FrequencyDefinitionBase"/>.
        /// </summary>
        public virtual uint DfDtScalingValue
        {
            get
            {
                return m_dfdtScale;
            }
            set
            {
                m_dfdtScale = value;
            }
        }

        /// <summary>
        /// Gets a <see cref="Dictionary{TKey,TValue}"/> of string based property names and values for this <see cref="FrequencyDefinitionBase"/> object.
        /// </summary>
        public override Dictionary<string, string> Attributes
        {
            get
            {
                Dictionary<string, string> baseAttributes = base.Attributes;

                baseAttributes.Add("df/dt Offset", DfDtOffset.ToString());
                baseAttributes.Add("df/dt Scaling Value", DfDtScalingValue.ToString());

                return baseAttributes;
            }
        }

        #endregion

        #region [ Methods ]

        /// <summary>
        /// Populates a <see cref="SerializationInfo"/> with the data needed to serialize the target object.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> to populate with data.</param>
        /// <param name="context">The destination <see cref="StreamingContext"/> for this serialization.</param>
        public override void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            base.GetObjectData(info, context);

            // Serialize frequency definition
            info.AddValue("dfdtScale", m_dfdtScale);
            info.AddValue("dfdtOffset", m_dfdtOffset);
        }

        #endregion
    }
}